include/revng/Pipeline/ContainerFactory.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/Pipeline/Container.h"
+#include "revng/Pipeline/SavableObject.h"
+
+namespace Pipeline {
+
+/// The pipeline system requires to be able to produce different copies of
+/// each Container, and thus the factories of non default constructible
+/// containers must be specified. class that extends this one must simply
+/// implement createEmpty
+///
+/// Unfortunatelly, not all container are default constructible, since
+/// they behaviour may depends on some external factor, as an example
+/// a llvm container requires a llvm context to be able to create a empty module
+class ContainerFactory {
+public:
+  virtual std::unique_ptr<ContainerBase> createEmpty() const = 0;
```

Comment:
Can't the whole `ContainerFactory` hierarchy be a `std::function<ContainerBase(void)>`?
Then in `registerDefaultConstructibleFactory` you can create the obvious lambda on the fly.
This approach could lead to entirely drop this file.

Reply:
the issue is that std::functions are not copiable, and thus they ruin all the objects that contain them

Reply:
Uhm...

> std::function satisfies the requirements of CopyConstructible and CopyAssignable.

Reply:
wait, how does it work? all lambdas are copiable?  i have to look into this